platform/ios/MGLUserLocationAnnotationView.m

```diff
@@ -48,164 +55,285 @@ - (instancetype)initWithCoder:(NSCoder *)decoder
-            _haloLayer = [CALayer layer];
-            _haloLayer.bounds = CGRectMake(0, 0, haloImage.size.width, haloImage.size.height);
-            _haloLayer.contents = (__bridge id)[haloImage CGImage];
-            _haloLayer.position = CGPointMake(super.layer.bounds.size.width / 2.0, super.layer.bounds.size.height / 2.0);
+            CGFloat headingIndicatorSize = MGLUserLocationAnnotationHaloSize;
             
-            [CATransaction begin];
+            _headingIndicatorLayer = [CALayer layer];
+            _headingIndicatorLayer.bounds = CGRectMake(0, 0, headingIndicatorSize, headingIndicatorSize);
+            _headingIndicatorLayer.position = CGPointMake(super.bounds.size.width / 2.0, super.bounds.size.height / 2.0);
+            _headingIndicatorLayer.contents = (__bridge id)[[self headingIndicatorTintedGradientImage] CGImage];
+            _headingIndicatorLayer.contentsGravity = kCAGravityBottom;
+            _headingIndicatorLayer.contentsScale = [UIScreen mainScreen].scale;
+            _headingIndicatorLayer.opacity = 0.4;
+            _headingIndicatorLayer.shouldRasterize = YES;
+            _headingIndicatorLayer.rasterizationScale = [UIScreen mainScreen].scale;
+            _headingIndicatorLayer.drawsAsynchronously = YES;
             
-            [CATransaction setAnimationDuration:3.5];
-            [CATransaction setAnimationTimingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut]];
-            
-            // scale out radially
-            //
-            CABasicAnimation *boundsAnimation = [CABasicAnimation animationWithKeyPath:@"transform"];
-            boundsAnimation.repeatCount = MAXFLOAT;
-            boundsAnimation.fromValue = [NSValue valueWithCATransform3D:CATransform3DMakeScale(0.1, 0.1, 1.0)];
-            boundsAnimation.toValue   = [NSValue valueWithCATransform3D:CATransform3DMakeScale(2.0, 2.0, 1.0)];
-            boundsAnimation.removedOnCompletion = NO;
+            [self.layer insertSublayer:_headingIndicatorLayer below:_dotBorderLayer];
+        }
+        
+        // heading indicator accuracy mask (fan-shaped)
+        //
+        if ( ! _headingIndicatorMaskLayer && self.annotation.heading.headingAccuracy)
+        {
+            _headingIndicatorMaskLayer = [CAShapeLayer layer];
+            _headingIndicatorMaskLayer.frame = _headingIndicatorLayer.bounds;
+            _headingIndicatorMaskLayer.path = [[self headingIndicatorClippingMask] CGPath];
+
+            // apply the mask to the halo-radius-sized gradient layer
+            _headingIndicatorLayer.mask = _headingIndicatorMaskLayer;
             
-            [_haloLayer addAnimation:boundsAnimation forKey:@"animateScale"];
+            _oldHeadingAccuracy = self.annotation.heading.headingAccuracy;
+        }
+        
+        // update accuracy ring
+        //
+        if (_accuracyRingLayer)
+        {
+            // FIX: This stops EVERYTHING... and that isn't necessarily necessary, now is it?
```

Comment:
I moved the zoom and heading accuracy is-necessary checks into the larger `if` statement in df2c1f5, which should be less jarring:

```diff
diff --git a/platform/ios/MGLUserLocationAnnotationView.m b/platform/ios/MGLUserLocationAnnotationView.m
index a6310a8..536326d 100644
--- a/platform/ios/MGLUserLocationAnnotationView.m
+++ b/platform/ios/MGLUserLocationAnnotationView.m
@@ -119,16 +119,10 @@ const CGFloat MGLUserLocationAnnotationHaloSize = 115.0;
             _oldHeadingAccuracy = self.annotation.heading.headingAccuracy;
         }
         
-        // update accuracy ring
+        // update accuracy ring (if zoom or horizontal accuracy have changed)
         //
-        if (_accuracyRingLayer)
+        if (_accuracyRingLayer && (_oldZoom != self.mapView.zoomLevel || _oldHorizontalAccuracy != self.annotation.location.horizontalAccuracy))
         {
-            // FIX: This stops EVERYTHING... and that isn't necessarily necessary, now is it?
-            if (_oldZoom == self.mapView.zoomLevel && _oldHorizontalAccuracy == self.annotation.location.horizontalAccuracy)
-            {
-                return;
-            }
-            
             CGFloat accuracyRingSize = [self calculateAccuracyRingSize];
             
             // only show the accuracy ring if it won't be obscured by the location dot
```